oeps/best-practices/oep-0018/decisions/0001-forbid-url-dependencies.rst

```diff
@@ -0,0 +1,287 @@
+  via VCS URL, whereas PyPI-hosted packages can be downloaded as pre-built wheels; and
+* when re-installing a dependencies, ``pip`` must fully re-install VCS-hosted
+  packages, whereas it will skip over any PyPI-hosted package whose version number has not
+  changed.
+
+We confirmed our theoretical understanding with a `cursory performance comparison`_.
+It demonstrated that installing a package into a fresh system from a VCS URL took
+nearly twice as a long as installing the same package from PyPI. It also
+demonstrated that installing an already-installed dependency from a VCS URL
+took four times as long as doing so for a package installed via PyPI.
+
+Upgrades
+--------
+
+OEP-18 describes a format for PyPI-hosted dependencies that allows for
+automated upgrades to be performed using the ``make uprade`` Makefile target.
+This target upgrades all version pins to the latest version, as allowed by the
+version restrictions specified in ``constraints.txt``.
+
+VCS URL dependencies work completely outside of this system. Each VCS URL
+is essentially a version pin, which cannot be upgraded by automated tooling.
+This runs counter to the "upgrade pins by default" philosophy of OEP-18.
+edx-platform's ``github.in`` file has VCS URL pins that have not been
+updated for over three years.
+
+Lack of reason to allow VCS URLs
+--------------------------------
+
+As noted above, OEP-18 carved out two exceptions in which VCS URLs were
+acceptable. Firstly:
+
+    You need to test a release candidate of the dependency to make sure it will
+    work with your code.
+
+This is a valid thing to test, but it should be done locally or in some sort of
+sandbox using a feature branch. A VCS URL should never be merged to main or master
+for testing purposes.
+
+Secondly:
+
+    You critically need a fix for a package which has not yet been included in
+    a release, and you cannot arrange for a release to be made in a timely
+    manner.
+
+Generally, fixes can be achieved by reverting to a previous, stable package version.
+In the rare event that reverting to a previous package version is not viable,
+Open edX repository maintainers can easily cut new releases of packages in order
+to accomodate a forward fix. If a repository maintainer is unresponsive, this
+issue can be escalated to tCRIL, which can cut an emergency release if necessary. If
+the repository is outside of the Open edX project, then the repository can be forked, and
+the package re-released to PyPI under a new name.
+
+Finally, even if those reasons were valid, they are *not* the reasons behind
+`the VCS URL dependencies installed in to edx-platform as of Nutmeg
+<https://github.com/openedx/edx-platform/blob/open-release/nutmeg.2/requirements/edx/github.in#L54-L77>`_
+Following ``git blame`` shows that the current VCS URL dependencies
+exist simply because it took fewer steps to set them up than it would have to
+install them from PyPI.
```

Comment:
A few other interesting use cases that have come up in the past:

* A dependency suddenly deletes all of its old PyPI releases, as in https://github.com/openedx/edx-platform/pull/27168 .  This removes the "downgrade" option and makes it critical to fix the dependency ASAP because it's breaking all deployments and installations.
* Something about codejail is preventing it from working correctly when installed from PyPI; Arbi-BOM has a new idea regarding what causes that and is currently looking into it, but it's not fixed yet.  https://openedx.atlassian.net/browse/BOM-2481
* We had to temporarily fork celery just to get useful logging data to diagnose a recurring problem in production: https://github.com/openedx/edx-platform/pull/25649
* Sometimes upstream drops previous LTS Django support before releasing a version with the new LTS Django support, like in https://github.com/openedx/edx-platform/pull/23257 .

I don't want to argue that installing from VCS is the only way to solve problems like these.  I just want to make sure that we have a replacement process that allows us to very quickly deploy a specific commit from a dependency before we eliminate this as an option.

Reply:
Good to know @jmbowman . I'm glad you saw this draft PR and brought these up before I went any further with this.

I believe the codejail issue was an [instance of an incorrect setup.py](https://github.com/openedx/codejail/pull/140). I don't have good alternative workflows for the other instances you brought up, though.

To be honest, I don't think BOM teams adding temporary URL dependencies in order to speed up upgrades and handle critical issues is the problem here. The problem is when URL dependencies are added as a quick-and-dirty way of getting a feature out the door. So, maybe the solution isn't to forbid VCS URLs altogether. Perhaps it'd be better to, for example, use CODEOWNERS to require review from one of the BOM teams when editing github.in, helping ensure the file isn't misused.

Reply:
I definitely agree that there are things in github.in that have stayed there far too long.  Like, there's never been a proper release of our drag & drop assessment XBlock.  But we may need something more like a repo health check and a followup process for remediation than an outright ban.